### PR TITLE
Add checks for account confusions

### PR DESCRIPTION
**Pull request by zatdeltafi**

### CODE REVIEW

File: solaudit/checkers.py [hunk 1]
```diff
@@ -35,5 +35,15 @@ def missingSignerCheckChecker(program: Program) -> None:
...
+    if len(program.pub_exprs) > 1:
+        print("====== Warning: potential account confusions are detected for below struct definitions =====")
+        for line, expr in program.pub_exprs.items():
+            print("line: %d %s" % (line, expr))
+            exprs[line] = expr
+    return exprs
 
-CHECKERS = [overUnderFlowChecker, missingSignerCheckChecker]
+CHECKERS = [overUnderFlowChecker,
+		missingSignerCheckChecker,
```
Comment (ypeng-deltafi) [inline solaudit/checkers.py:48]: can you use space instead of tab?

File: solaudit/checkers.py [hunk 2]
```diff
@@ -35,5 +35,15 @@ def missingSignerCheckChecker(program: Program) -> None:
 
     return func_missing_signer_check
 
+def accountConfusionsChecker(program: Program) -> map:
+    exprs = {}
+    if len(program.pub_exprs) > 1:
+        print("====== Warning: potential account confusions are detected for below struct definitions =====")
+        for line, expr in program.pub_exprs.items():
+            print("line: %d %s" % (line, expr))
+            exprs[line] = expr
```
Comment (ypeng-deltafi) [inline solaudit/checkers.py:44]: it seems expr is equal to program.pub_exprs. why not just return program.pub_exprs directly?

File: solaudit/checkers.py [hunk 3]
```diff
@@ -35,5 +35,15 @@ def missingSignerCheckChecker(program: Program) -> None:
 
     return func_missing_signer_check
 
+def accountConfusionsChecker(program: Program) -> map:
```
Comment (ypeng-deltafi) [inline solaudit/checkers.py:38]: I think in python the conversion is use camel_function_name

Comment (ypeng-deltafi) [reply solaudit/checkers.py:38]: do we want to use https://google.github.io/styleguide/pyguide.html?

File: solaudit/models.py [hunk 4]
```diff
@@ -21,9 +21,11 @@ def __init__(self, name) -> None:
...
 
 class Program:
+
     def __init__(self) -> None:
         self.algbra_exprs = {}
         self.functions = {}
+        self.pub_exprs = {}
         pass
 
     def handle_function_def(self, s: str, loc: int, tokens: ParseResults) -> None:
```
Comment (ypeng-deltafi) [inline solaudit/models.py:31]: this one use camel. Can we make the checkers.py consistent with this>

File: solaudit/models.py [hunk 5]
```diff
@@ -48,3 +50,5 @@ def handle_assignment_stat(self, s: str, loc: int, tokens: ParseResults) -> None
         # Handle cases like += -=.
         if "operator" in tokens:
             self.algbra_exprs[lineno(loc, s)] = line(loc, s)
+    def handle_struct_stat(self, s: str, loc: int, tokens: ParseResults) -> None:
```
Comment (ypeng-deltafi) [inline solaudit/models.py:53]: add new line between functions

File: solaudit/parsers.py [hunk 6]
```diff
@@ -89,12 +90,38 @@ def getProgramParser(program: Program) -> pp.ParserElement:
     )
     return_stat = RETURN + exp + SEMI
 
+    pubkey_stat = pp.Literal("Pubkey")
```
Comment (ypeng-deltafi) [inline solaudit/parsers.py:93]: what is pp? can we use better name?

File: solaudit/parsers.py [hunk 7]
```diff
@@ -89,12 +90,38 @@ def getProgramParser(program: Program) -> pp.ParserElement:
...
+    struct_member_pubkey_expr = (
+        pp.Optional(PUB)
+        + var
+        + pp.Group(COLON + pubkey_stat + COMMA)
```
Comment (ypeng-deltafi) [inline solaudit/parsers.py:97]: what does this expression mean? can you add some comments?

File: solaudit/parsers.py [hunk 8]
```diff
@@ -89,12 +90,38 @@ def getProgramParser(program: Program) -> pp.ParserElement:
...
+    )
+
+    struct_member_any_expr = (
+        pp.Optional(PUB)
+        + var
+        + pp.Group(COLON + pp.Word(pp.alphanums) + COMMA)
+    )
+
+    struct_stat = (
```
Comment (JeremyLi28) [inline solaudit/parsers.py:106]: Shall we check if the account has a `type` defined before Pubkey? the parsing will fail if there are any fields defined before pubkey.

File: solaudit/checkers.py (hunk 2 above)
Comment (zatdeltafi) [reply solaudit/checkers.py:44]: expr is a part member of program.pub_exprs

File: solaudit/checkers.py (hunk 3 above)
Comment (zatdeltafi) [reply solaudit/checkers.py:38]: yes, this is good suggestion. We need to follow this guide.

File: solaudit/parsers.py (hunk 6 above)
Comment (zatdeltafi) [reply solaudit/parsers.py:93]: This is "import pyparsing as pp" from the first line.

File: solaudit/parsers.py (hunk 8 above)
Comment (zatdeltafi) [reply solaudit/parsers.py:106]: Most cases of account confusion happen if there are two structures, the first member of each structure is an account pubkey. So this struct model is just used to catch this case. 
We may add additional type before pubkey to detect more complex cases in the future.

File: solaudit/parsers.py (hunk 7 above)
Comment (zatdeltafi) [reply solaudit/parsers.py:97]: added comments

**Review by ypeng-deltafi — APPROVED** [review]

**Pull request merged**